### DISCUSSION TITLE
Renderer as a child inside another window

### DISCUSSION THREAD

**Discussion by andreiltd** [marc2332/freya#516] [Q&A]
Hello :wave: 

Is it possible to integrate Freya for UI development within an application that is built on winit, similar to the way `wry` allows the creation of [child webviews](https://docs.rs/wry/latest/wry/#child-webviews)? Specifically, I'm interested in understanding if Freya can be used to design 2D menus or interfaces within a game application which owns the window and event loop.

**Comment by marc2332** [comment]
Hey Tomasz!

I think it could be possible, although I have never attempted anything as that. Freya depends on some Skia APIs, so it's not exactly graphics-agnostic, perhaps that is not a problem.

Freya at its core is event loop-agnostic, so it should be possible to integrate it into another winit app. It even has a headless testing runner that simulates an event loop and platform inputs (like cursor movement, and that stuff), so, my point is that it would only need to be integrated properly. Perhaps getting skia to work on top is the most difficult part? I am not sure

**Comment by felipetesc** [comment]
Hello. I'm just guessing here, but I think, child webviews, alert dialogs, and the such, are actually something related to winit itself(as you stated) and how to integrate to Freya.  Here we can find an example of multi windows: https://github.com/rust-windowing/winit/blob/master/examples/multiwindow.rs. So, perhaps, if there is high level way to coordinate multi windows manageament, using for example, use_state, and use_signal, and [the event bubbling](https://github.com/marc2332/freya/pull/268), it will be possible. About  the menubar, this Rust crate, could help https://lib.rs/crates/menubar, although I donno the current state of it.

**Reply by marc2332** [reply]
Hey. Tomasz isn't refering to multiwindows, he is refering on how to render Freya on top of a videogame. There is no need for any multi window management or any tweak to the state or events system (I think?). It all comes down to having two apps running under the same event loop and two renderers rendering in the same window. For instance, he refers to child webviews from wry, these are not child windows, but child webviews running inside the same window

**Reply by felipetesc** [reply]
I see, he wants something like egui but ~with html for GUI (removed)~ with Freya . 
What about this project: https://github.com/DioxusLabs/blitz ?

PS: I don't now if we would need skia, though ...

**Reply by marc2332** [reply]
No, there is no HTML, wry was simply an example on how they integrate multiple GUIs in the same window.

**Reply by andreiltd** [reply]
I did some initial research and it should be doable via Skia by setting translation matrix on canvas:

```rust
let matrix = Matrix::translate(Vector::new(100.0, 100.0));
canvas.set_matrix(&matrix);
```

which causes the coordinate system to be shifted such that the point (100, 100) in the original coordinate system is now treated as the origin (0, 0) so any drawing will be shifted as well. This would, hopefully, allow to create subregion for drawing.

**Reply by marc2332** [reply]
Sorry for late reply. Yes, that would move the rendering wherever you want, but it would break the event system as freya would still think the elements are wherever they were in the first place, so events would also need to be translated to the new coordinates

**Reply by felipetesc** [reply]
With this event system changes, and more tweaks,  would be possible to use Freya with webgl (wasm),  Fyrox, SDL Canvas, and so own ?

**Reply by marc2332** [reply]
Everything is possible with enough tweaks, Freya could run in wasm "easily" as Skia supports web, but,running Freya under Fyrox or other renderers would require making Freya renderer-agnostic, which isn't currently the case as it's "highly" integrated with Skia
